### PR TITLE
Ethereum api and big.Int refactor

### PR DESCRIPTION
**Pull request by sasurobert**
Ethereum api and big.Int refactor.
Slight changes of core api, as cgo build is problematic.

### CODE REVIEW

File: arwen/context/bigIntContainer.go [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+func (b *bigIntContainer) Clean() {
+	b.mutex.Lock()
+	b.mappedValues = make(map[int32]*big.Int)
+	b.mutex.Unlock()
+}
+
+func (b *bigIntContainer) Put(value int64) int32 {
+	b.mutex.Lock()
+	b.mutex.Unlock()
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply arwen/context/bigIntContainer.go]: done

File: arwen/ethapi/ethereumei.go [hunk 2]
```diff
@@ -0,0 +1,543 @@
...
+// extern void getTxGasPrice(void *context, i32ptr valueOffset);
+// extern void logTopics(void *context, i32ptr dataOffset, i32 length, i32 numberOftopics, i32ptr topic1, i32ptr topic2, i32ptr topic3, i32ptr topic4);
+// extern long long getBlockNumber(void *context);
+// extern void getTxOrigin(void *context, i32ptr resultOffset);
+// extern void finish(void *context, i32ptr dataOffset, i32 length);
+// extern void revert(void *context, i32ptr dataOffset, i32 length);
+// extern i32 getReturnDataSize(void *context);
+// extern void returnDataCopy(void *context, i32ptr resultOffset, i32 dataOffset, i32 length);
+// extern void selfDestruct(void *context, i32ptr addressOffset);
+// extern long long getBlockTimestamp(void *context);
```
Comment (camilbancioiu) [inline arwen/ethapi/ethereumei.go]: Some of these CGO exports seem to be conflicting with the ones from ElrondEI. This branch does not build on my machine:
```
multiple definition of `getExternalBalance' ... (etc)
multiple definition of `storageStore' ... (etc)
multiple definition of `storageLoad' ... (etc)
multiple definition of `getCaller' ... (etc)
multiple definition of `getGasLeft' ... (etc)
multiple definition of `getBlockTimestamp' ... (etc)
```

File: arwen/interface.go [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+	GetVMInput() vmcommon.VMInput
+	Transfer(destination []byte, sender []byte, value *big.Int, input []byte, gas int64) (gasLeft int64, err error)
+	GetStorage(addr []byte, key []byte) []byte
+	SetStorage(addr []byte, key []byte, value []byte) int32
+	GetCode(addr []byte) []byte
+	GetCodeSize(addr []byte) int32
+	SignalUserError()
+	Finish(data []byte)
+	WriteLog(addr []byte, topics [][]byte, data []byte)
+	SelfDestruct(addr []byte, beneficiary []byte)
```
Comment (camilbancioiu) [inline arwen/interface.go]: For readability, sort these lines alphabetically, to highlight common methods with the `HostContext` interface.

File: arwen/interface.go [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+	SetStorage(addr []byte, key []byte, value []byte) int32
+	GetBalance(addr []byte) []byte
+	BlockHash(nonce int64) []byte
+	GetVMInput() vmcommon.VMInput
+	GetSCAddress() []byte
+	WriteLog(addr []byte, topics [][]byte, data []byte)
+	Transfer(destination []byte, sender []byte, value *big.Int, input []byte, gas int64) (gasLeft int64, err error)
+	Finish(data []byte)
+	BlockChainHook() vmcommon.BlockchainHook
+	SignalUserError()
```
Comment (camilbancioiu) [inline arwen/interface.go]: For readability, sort these lines alphabetically, to highlight common methods with the `EthContext` interface.

File: arwen/context/arwen.go [hunk 5]
```diff
@@ -539,3 +529,55 @@ func (host *vmContext) Transfer(destination []byte, sender []byte, value *big.In
...
+		if err != nil {
+			return nil
+		}
+
+		newInput = append(newInput, methodSelectors[0:4]...)
+	}
+
+	for _, arg := range host.vmInput.Arguments {
+		currInput := make([]byte, arwen.HashLen)
+		copy(currInput[arwen.HashLen-len(arg.Bytes()):], arg.Bytes())
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply arwen/context/arwen.go]: yes. That is the way ethereum wants the input.

File: arwen/context/bigIntContainer.go [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+	return newIndex
+}
+
+func (b *bigIntContainer) GetOne(id int32) *big.Int {
+	b.mutex.Lock()
+	defer b.mutex.Unlock()
+
+	if _, ok := b.mappedValues[id]; !ok {
+		b.mappedValues[id] = big.NewInt(0)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply arwen/context/bigIntContainer.go]: I changed the put function. It searches for the closest new open address.

File: arwen/elrondapi/bigintOps.go [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+
+//export bigIntGetBytes
+func bigIntGetBytes(context unsafe.Pointer, reference int32, byteOffset int32) int32 {
+	instCtx := wasmer.IntoInstanceContext(context)
+	hostContext := arwen.GetBigIntContext(instCtx.Data())
+
+	bytes := hostContext.GetOne(reference).Bytes()
+
+	err := arwen.StoreBytes(instCtx.Memory(), byteOffset, bytes)
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply arwen/elrondapi/bigintOps.go]: deleted. nothing happens.

File: arwen/elrondapi/elrondei.go [hunk 8]
```diff
@@ -0,0 +1,390 @@
...
+	//TODO implement
+	return 10000
+}
+
+//export getOwner
+func getOwner(context unsafe.Pointer, resultOffset int32) {
+	instCtx := wasmer.IntoInstanceContext(context)
+	hostContext := arwen.GetErdContext(instCtx.Data())
+
+	owner := hostContext.GetSCAddress()
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply arwen/elrondapi/elrondei.go]: the owner of the smart contract address, is the smart contract. I do not know if they want to know the account which created it.

File: arwen/context/errors.go [hunk 9]
```diff
@@ -0,0 +1,5 @@
+package context
+
+import "errors"
+
+var ErrInitFuncCalledInRun = errors.New("it is not allowed to call init in run")
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply arwen/context/errors.go]: done

File: arwen/elrondapi/elrondei.go [hunk 10]
```diff
@@ -0,0 +1,390 @@
...
+	if err != nil {
+	}
+}
+
+//export blockHash
+func blockHash(context unsafe.Pointer, nonce int64, resultOffset int32) int32 {
+	instCtx := wasmer.IntoInstanceContext(context)
+	hostContext := arwen.GetErdContext(instCtx.Data())
+
+	hash := hostContext.BlockHash(nonce)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply arwen/elrondapi/elrondei.go]: good observation. will change blockchain hook implementation.
added TODO

File: arwen/ethapi/ethereumei.go (hunk 2 above)
Comment (sasurobert) [reply arwen/ethapi/ethereumei.go]: done

**Pull request merged**